### PR TITLE
[Hyper-cli] embed hpm

### PR DESCRIPTION
**Pull request by chabou**
This PR has PR #2342 already merged into it.

Next steps:
* [X] Prevent `Hyper CLI` to create a new instance of `Hyper` if it is already launched. It should act the same way that launching Hyper twice by icon.
  * [X] macOS (new tab)
  * [X] Windows (new window)
  * [X] Linux (new window)
* [X] Test it on Linux
* [X] Test it on Windows
  * [X] Cmd
  * [X] PowerShell
  * [X] GitBash/CygWin
  * [X] WSL ⚠️  we can only launch `Hyper` and not install/uninstall/list plugins due to Env var problem: see  https://github.com/Microsoft/WSL/issues/1494

Bonus:
* [X]  `hyper <path>` should open Hyper in this path
  * [X] macOS
  * [X] Windows
  * [X] Linux
  * [X] Check if path exists

Screencast: 
hyper-cli

### CODE REVIEW

**Comment by chabou** [comment]
There is an error on Linux
```
Failed to symlink Hyper CLI { Error: EACCES: permission denied, unlink '/usr/local/bin/hyper'
  errno: -13,
  code: 'EACCES',
  syscall: 'unlink',
  path: '/usr/local/bin/hyper' }
```

```
parallels@ubuntu:~/hyper$ ls -lah /usr/local/bin/
total 8.0K
drwxr-xr-x  2 root root 4.0K Oct 23 00:02 .
drwxr-xr-x 10 root root 4.0K Jul 19  2016 ..
lrwxrwxrwx  1 root root   16 Oct 23 00:02 hyper -> /opt/Hyper/hyper
```

**Comment by chabou** [comment]
One cool side effect is that Hyper CLI is our `open <path> in new tab` feature (on macOS):
hyper-cli-openinnewtab


But the more important thing is: THIS PR IS READY TO MERGE 🎉 🎉  🎉

**Comment by peterfraedrich** [comment]
I've submitted a PR for `hpm` that deprecates the use of `npm-name` and adds authenticated proxy support. 
[Link To PR](https://github.com/zeit/hpm/pull/100)

**Comment by Stanzilla** [comment]
I can't add reviewers but I nominate @albinekb :)

**Comment by htkoca** [comment]
Instead of going to the path in the first argument, can it be a command? hyper `cd /path` rather than just `/path` . This will allow `cd` to handle path existence (which it is designed to do), and also allow other applications to run things in hyper (such as an SFTP program opening a SSH connection in hyper)

File: cli/index.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
+// This is a CLI tool, using console is OK
+/* eslint no-console: 0 */
+const {spawn, exec} = require('child_process');
+const {isAbsolute, resolve} = require('path');
+const {existsSync} = require('fs');
+const pify = require('pify');
+const args = require('args');
+const chalk = require('chalk');
+const opn = require('opn');
+// const columnify = require('columnify');
```
Comment (albinekb) [inline cli/index.js]: remove

File: cli/index.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+// const ora = require('ora');
+const api = require('./api');
+
+let commandPromise;
+
+const checkConfig = () => {
+  if (api.exists()) {
+    return true;
+  }
+  let msg = chalk.red(`Error! Config file not found: ${api.configPath}\n`);
```
Comment (albinekb) [inline cli/index.js]:
```js

Comment (albinekb) [inline cli/index.js]:
```const

Comment (albinekb) [inline cli/index.js]:
```

File: cli/index.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  msg += 'Please launch Hyper and retry.';
```
Comment (albinekb) [inline cli/index.js]:
```js
const message = `
${chalk.red(`Error! Config file not found: ${api.configPath}`)}
Please launch Hyper and retry.
`
```

File: app/utils/cli-install.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+exports.addBinToUserPath = () => {
+  // Can't use pify because of param order of Registry.values callback
+  return new Promise((resolve, reject) => {
+    const envKey = new Registry({hive: 'HKCU', key: '\\Environment'});
+    envKey.values((err, items) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+      // C:\Users\<user>\AppData\Local\hyper\app-2.0.4\resources\bin
```
Comment (albinekb) [inline app/utils/cli-install.js]: `// C:\Users\<user>\AppData\Local\hyper\app-<version>\resources\bin`

File: cli/api.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
+const fs = require('fs');
+const os = require('os');
+const npmName = () => Promise.resolve(); //require('npm-name');
```
Comment (albinekb) [inline cli/api.js]: 🤔

File: cli/index.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+  if (plugins) {
+    console.log(plugins);
+  } else {
+    console.log(chalk.red(`No plugins installed yet.`));
+  }
+  process.exit(1);
+});
+/*
+const lsRemote = () => {
```
Comment (albinekb) [inline cli/index.js]: Remove?

File: cli/index.js (hunk 3 above)
Comment (albinekb) [reply cli/index.js]: Should we add an issue to make the cli ask to create the file if it does not exists?

File: cli/index.js (hunk 6 above)
Comment (chabou) [reply cli/index.js]: It will be reactivated when NPM API request (URL and criteria) will be chosen.

File: app/utils/cli-install.js (hunk 4 above)
Comment (chabou) [reply app/utils/cli-install.js]: I'll fix this

File: cli/index.js (hunk 1 above)
Comment (chabou) [reply cli/index.js]: It will be used by lsRemote

File: cli/index.js (hunk 3 above)
Comment (chabou) [reply cli/index.js]: Yes. 
Config creation should remain an `Hyper` task. And yes, we should improve this automatic creation. But this is not a blocker. Most of case, config file will be present.

Comment (albinekb) [reply cli/index.js]: 👌

File: cli/api.js (hunk 5 above)
Comment (chabou) [reply cli/api.js]: I hoped that this PR would have been merged sooner: https://github.com/dominictarr/rc/pull/94

**Comment by chabou** [comment]
@htkoca thank you for your feedback. You're right. Passing command could be very usefull.

Imo, I prefer command as an option like your `--run` suggestion. Most of time, users will only want to open `Hyper` on a (default) location. I don't want to oblige them to add `cd` to their path.

@albinekb @rauchg @leo  Thoughts ?

File: cli/index.js (hunk 1 above)
Comment (chabou) [reply cli/index.js]: fixed

File: app/utils/cli-install.js (hunk 4 above)
Comment (chabou) [reply app/utils/cli-install.js]: fixed

File: cli/api.js (hunk 5 above)
Comment (chabou) [reply cli/api.js]: fixed

File: cli/index.js (hunk 6 above)
Comment (chabou) [reply cli/index.js]: fixed

**Comment by chabou** [comment]
For `install`, `unsinstall`, `docs` commands, `hyper-` prefix is added if not present.
For `search` and `ls-remote` commands, results are filtered to show only `hyper-` prefixed packages.

File: build/win/hyper [hunk 7]
```diff
@@ -5,6 +5,9 @@ NAME="Hyper"
 HYPER_PATH="$(dirname "$(dirname "$(dirname "$(realpath "$0")")")")"
 ELECTRON="$HYPER_PATH/$NAME.exe"
 if grep -q Microsoft /proc/version; then
+	echo "Warning! Due to WSL limitations, you can use CLI commands here. Please use Hyper CLI on cmd, PowerShell or GitBash/CygWin."
```
Comment (amckinlay) [inline build/win/hyper]: What does this mean? Do you mean "you cannot use Hyper CLI here"?

**Comment by amckinlay** [comment]
@chabou I left a comment about the language of the warning message in `build/win/hyper`.

As an aside, I wonder why my comment was on an outdated commit.

File: build/win/hyper (hunk 7 above)
Comment (Stanzilla) [reply build/win/hyper]: yeah just missing the `'t`

**Comment by chabou** [comment]
Yes this is a typo.
Can you send a PR or open an issue if you can't?
Thank you 🙏

**Pull request merged**